### PR TITLE
Patchouli Book

### PR DESCRIPTION
**Pull request by thiakil**
Very WIP Patchouli book.

Help welcomed.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -77,6 +80,12 @@ sourceSets {
     datagentools { setupDataGenSourceSet(sourceSets.datagentools, sourceSets.tools, 'tools') }
 }
 
+// no kotlin for the main mod
+sourceSets.main.kotlin.srcDirs.removeAll()
+sourceSets.additions.kotlin.srcDirs.removeAll()
+sourceSets.defense.kotlin.srcDirs.removeAll()
+sourceSets.tools.kotlin.srcDirs.removeAll()
```
Comment (pupnewfster) [inline build.gradle:87]: Shouldn't this also be getting removed from the api and generator modules?

File: src/api/java/mekanism/api/providers/IBlockProvider.java [hunk 2]
```diff
@@ -22,12 +23,14 @@ default boolean blockMatches(Block other) {
     }
 
     @Override
+    @Nonnull
```
Comment (pupnewfster) [inline src/api/java/mekanism/api/providers/IBlockProvider.java:26]: Personally I am more of a fan of putting the Nonnull above the override as it is a shorter annotation

**Comment by Oct0bass** [comment]
Nice!  This looks really interesting already.

File: src/datagen/main/kotlin/mekanism/patchouli/MekanismMainBookProvider.kt [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+                                    "\$(li)If you want to maintain your altitude, choose Hover mode. If you want to ascend/descend rapidly, use Regular mode. If you want to conserve fuel while trekking across hills, mountains, consider Disabled mode."+
+                                    "\$(li)The Jetpack can be paired with the Free Runners to protect against fall damage."
+                        }
+                    }
+                    ATOMIC_DISASSEMBLER("The Atomic Disassembler is Mekanism's an all-in-one tool, essentially the ultimate, electronic version of the Paxel (working at any mining level). Also functions as a Hoe & Scoop (Forestry)$(p)The Atomic Disassembler has multiple modes that can be cycled with $(k:sneak) + right click.") {
+                        text {
+                            title = "Normal Mode"
+                            text = "Base speed setting, single block.$(li)Roughly equivalent to Efficiency II.$(li)Right click Dirt to till a 3x3 area to Farmland$(li)Right click grass to make a 3x3 Grass Path, right click again to till"
```
Comment (pupnewfster) [inline src/datagen/main/kotlin/mekanism/patchouli/MekanismMainBookProvider.kt:71]: This isn't actually true anymore the right click to till, make grass paths, and strip logs got moved to the meka tool's farming module.

Comment (thiakil) [reply src/datagen/main/kotlin/mekanism/patchouli/MekanismMainBookProvider.kt:71]: Ah, good spotting. I've mostly been copying from the wiki

File: build.gradle [hunk 4]
```diff
@@ -273,6 +297,9 @@ minecraft {
                     sources(sourceSets.tools, sourceSets.datagentools)
                 }
             }
+
+            //For Mixin in other mods / Patchouli
+            property 'mixin.env.disableRefMap', 'true'
```
Comment (pupnewfster) [inline build.gradle:302]: This can be removed because we already have it above the mods block for datagen

Comment (thiakil) [reply build.gradle:302]: Well, no, this is not datagen... Patchouli is added as a mod for testing purposes

Comment (pupnewfster) [reply build.gradle:302]: no I mean the datagen block
```
data {

}
```

Comment (thiakil) [reply build.gradle:302]: Doh, sorry, blame mobile github

File: build.gradle [hunk 5]
```diff
@@ -616,6 +648,10 @@ compileJava {
     ]
 }
 
+tasks.withType(JavaCompile) {
+    options.encoding = 'UTF-8'
+}
```
Comment (pupnewfster) [inline build.gradle:653]: Unneeded it is like 10 lines above this.

File: build.gradle [hunk 6]
```diff
@@ -363,6 +380,12 @@ repositories {
             includeGroup 'appeng'
         }
     }
+    maven {
+        url 'https://maven.blamejared.com'
+        content {
+            includeGroup 'vazkii.patchouli'
+        }
+    }
```
Comment (pupnewfster) [inline build.gradle:388]: The includeGroup should just be moved into the existing one for jared's maven with CrT and JEITweaker

File: build.gradle [hunk 7]
```diff
@@ -438,6 +465,11 @@ dependencies {
     datagenmainImplementation fg.deobf("curse.maven:i-like-wood-biomes-o-plenty-437203:${ilikewood_bop_id}")
     datagenmainImplementation fg.deobf("curse.maven:i-like-wood-oh-the-biomes-youll-go-475110:${ilikewood_byg_id}")
     datagenmainRuntimeOnly fg.deobf("curse.maven:projecte-226410:${projecte_id}")
+
+    datagenmainCompile "org.jetbrains.kotlin:kotlin-stdlib-jdk8:$kotlin_version"
+    datagengeneratorsCompile "org.jetbrains.kotlin:kotlin-stdlib-jdk8:$kotlin_version"
+    datagenadditionsCompile "org.jetbrains.kotlin:kotlin-stdlib-jdk8:$kotlin_version"
+    datagentoolsCompile "org.jetbrains.kotlin:kotlin-stdlib-jdk8:$kotlin_version"
```
Comment (pupnewfster) [inline build.gradle:472]: strictly speaking this should also probably be added to datagendefenseCompile even if it is just a dummy module currently

File: build.gradle [hunk 8]
```diff
@@ -886,4 +922,14 @@ if (System.getenv("CURSEFORGE_KEY") != null || project.hasProperty('curseforgeKe
             addGameVersion "${minecraft_version}"
         }
     }
+}
+compileKotlin {
+    kotlinOptions {
+        jvmTarget = "1.8"
+    }
+}
```
Comment (pupnewfster) [inline build.gradle:930]: Would be nicer if there was a space before and after this block.

File: build.gradle (hunk 5 above)
Comment (thiakil) [reply build.gradle:653]: I did wonder about that, must be a merge error

**Comment by AdamSacha** [comment]
Hello, is this still being worked on?

**Comment by thiakil** [comment]
It's still on the list yeah, but motivation is hard

File: build.gradle [hunk 9]
```diff
@@ -121,6 +132,12 @@ def setupDataGenSourceSet(SourceSet datagenSourceSet, SourceSet baseSourceSet, S
     }
 }
 
+tasks.withType(org.jetbrains.kotlin.gradle.tasks.KotlinCompile).all {
+    kotlinOptions {
+        jvmTarget = "1.8"
```
Comment (pupnewfster) [inline build.gradle:137]: Should this be targetting java 17 now?

Comment (thiakil) [reply build.gradle:137]: Yes, is wip

**Pull request closed without merging**